### ISSUE TITLE
New Rule: Optional Property Write

### ISSUE THREAD

**Issue by Dschee** [realm/SwiftLint#1519]
This is a rule request. I'm willing to implement this myself and send a PR, but before doing that I'd like to know if the rule is wanted or not. To explain the desired rule, I'd like to first post the **custom rule** we are already using in our projects:

```yaml
optional_property_write:
    included: ".*.swift"
    regex: '\S+!\.\S+\s*=[^=]\s*'
    name: "Writing Properties on Optionals"
    message: "Don't use ! to write into properties on optionals – use ? instead."
    severity: warning
```

I felt like this rule might be useful for other projects as well, which is why I'm posting it. What do you think about it? For us it helps a lot to prevent common issues and enforce a shared style.

**Rationale**

When assigning a value to a property of a type, some people used the `!` syntax which will lead to a crash. In nearly all cases it is better to use `?` instead to ensure the app doesn't crash due to this single property assignment. In cases where the existence of the variable is required, the `if let` or `guard let` syntax should be used anyways.

A few examples:

```swift
let label: UILabel?

// ❌ not acceptable
label!.text = "someString"

// ✅ acceptable
label?.text = "someString"
```

**Comment by kimdv** [comment]
Is this not something similar to `force_upwrapping` rule?

**Comment by Dschee** [comment]
It is, but we are not using `force_unwrapping` since that one also works for read-only force unwrapping which we use in many places due to Storyboard and other APIs responding with optionals. This one could also be called `write_only_force_unwrapping` – it only works when trying to write by using force-unwrapping.

**Comment by Dschee** [comment]
Maybe this could also be implemented by adding a configuration option like `write_only` that defaults to `false` to keep the current behavior. But when `write_only` is set to true, only on write requests a warning (or error) will be shown. What do you think?
